framework/Models/HybridModel.py

```diff
@@ -0,0 +1,621 @@
+    converged = True
+    for targetName, metricInfo in outputDict.items():
+      for metricName, metricValues in metricInfo.items():
+        if max(metricValues) > self.romConvergence:
+          self.raiseADebug("maximum error: ", max(metricValues), " is greater than the given tolerance")
+          converged = False
+    return converged
+
+  def checkRomConvergence(self):
+    """
+      This function will check the convergence of all ROMs
+      @ In, None
+      @ Out, bool, True if all ROMs are converged
+    """
+    converged = True
+    for romInfo in self.romsDictionary.values():
+      if not romInfo['Converged']:
+        converged = False
+    if converged:
+      self.raiseADebug("All ROMs are converged")
+    return converged
+
+  def checkRomValidity(self, kwargs):
+    """
+      This function will check the validity of all roms
+      @ In, kwargs, dict,  is a dictionary that contains the information coming from the sampler,
+        a mandatory key is the sampledVars'that contains a dictionary {'name variable':value}
+      @ Out, None
+    """
+    allValid = False
+    if self.modelSelection == 'CrowdingDistance':
+      allValid = self.crowdingDistanceMethod(kwargs)
+    else:
+      self.raiseAnError(IOError, "Unknown model selection method ", self.modelSelection, " is given!")
+
+    if allValid:
+      self.raiseADebug("ROMs  are all valid for given model ", self.modelInstance.name)
+
+    return allValid
+
+  def crowdingDistanceMethod(self, kwargs):
+    """
+      This function will check the validity of all roms based on the crowding distance method
+      @ In, kwargs, dict,  is a dictionary that contains the information coming from the sampler,
+        a mandatory key is the sampledVars'that contains a dictionary {'name variable':value}
+      @ Out, None
+    """
+    allValid = True
+    for romInfo in self.romsDictionary.values():
+      valid = False
+      # generate the data for input parameters
+      paramsList = romInfo['Instance'].getInitParams()['Features']
+      trainInput = np.asarray(self._extractInputs(romInfo['Instance'].trainingSet, paramsList).values())
+      currentInput = np.asarray(self._extractInputs(kwargs['SampledVars'], paramsList).values())
+      coeffCD = self.computeCDCoefficient(trainInput, currentInput)
+      self.raiseADebug("Crowding Distance Coefficient: ", coeffCD)
+      if coeffCD > 0.2:
+        valid = True
+      romInfo['Valid'] = valid
+      if valid:
+        self.raiseADebug("ROM ",romInfo['Instance'].name, " is valid")
+      else:
+        allValid = False
+
+    return allValid
+
+  def _extractInputs(self,dataIn, paramsList):
+    """
+      Extract the the parameters in the paramsList from the given data object dataIn
+      @ dataIn, Instance or Dict, data object or dictionary contains the input and output parameters
+      @ paramsList, List, List of parameter names
+      @ localInput, dict,
+    """
+    localInput = dict.fromkeys(paramsList, None)
```

Comment:
I think this creates a link.... (pointer)

Reply:
what I found is:
The method fromkeys() creates a new dictionary with keys from seq and values set to value. Why is it a pointer?